support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,193 @@
+#include <condition_variable>
+#include <mutex>
+
+namespace snapchat::djinni {
+
+template <typename T>
+struct ValueHolder {
+    using type = T;
+    std::optional<T> value;
+    T getValueUnsafe() const {return *value;}
+};
+template <>
+struct ValueHolder<void> {
+    using type = bool;
+    std::optional<bool> value;
+    void getValueUnsafe() const {}
+};
+
+template<typename T>
+struct SharedState: ValueHolder<T> {
+    std::condition_variable cv;
+    std::mutex mutex;
+    std::exception_ptr exception;
+    std::function<void(std::shared_ptr<SharedState>)> handler;
+
+    bool isReady() const {
+        return this->value.has_value() || exception != nullptr;
+    }
+};
+
+template<typename T>
+using SharedStatePtr = std::shared_ptr<SharedState<T>>;
+
+template <typename T>
+class Future;
+
+template <typename T>
+class PromiseBase {
+public:
+    Future<T> getFuture();
+
+protected:
+    void setValue(typename ValueHolder<T>::type val) {
+        SharedStatePtr<T> sharedState;
+        {
+            std::unique_lock lk(_sharedState->mutex);
+            _sharedState->value = val;
+            sharedState = std::move(_sharedState);
+        }
+        callResultHandler(sharedState);
+    }
+
+    template <typename E>
+    void setException(const E& ex) {
+        setException(std::make_exception_ptr(ex));
+    }
+
+    void setException(std::exception_ptr ex) {
+        SharedStatePtr<T> sharedState;
+        {
+            std::unique_lock lk(_sharedState->mutex);
+            _sharedState->exception = ex;
+            sharedState = std::move(_sharedState);
+        }
+        callResultHandler(sharedState);
+    }
+private:
+    SharedStatePtr<T> _sharedState = std::make_shared<SharedState<T>>();
+
+    void callResultHandler(SharedStatePtr<T> sharedState) {
+        if (sharedState->handler) {
+            // handler already assigned
+            sharedState->handler(sharedState);
+        } else {
+            sharedState->cv.notify_all();
+        }
+    }
+};
+
+template <typename T>
+class Promise: public PromiseBase<T> {
+public:
+    using PromiseBase<T>::setValue;
+    using PromiseBase<T>::setException;
+};
+
+template <>
+class Promise<void>: public PromiseBase<void> {
+public:
+    void setValue() {
+        PromiseBase<void>::setValue(true);
+    }
+    using PromiseBase<void>::setException;
+};
+
+template<typename T>
+class Future {
+    template<typename U>
+    friend class PromiseBase;
+    
+    Future(SharedStatePtr<T> sharedState) : _sharedState(sharedState) {}
+public:
+    Future(const Future& other) = default;
+    Future(Future&& other) noexcept = default;
+    Future& operator= (Future&& other) noexcept = default;
+
+    bool isValid() const {
+        return _sharedState != nullptr;
+    }
+    
+    bool isReady() const {
+        std::unique_lock lk(_sharedState->mutex);
+        return _sharedState->isReady();
+    }
+    
+    void wait() const {
+        std::unique_lock lk(_sharedState->mutex);
```

Comment:
so, this locks the object until wait() is complete.   if someone calls isReady() while another thread is calling wait(), won't this cause isReady to hang until wait() is complete?

Reply:
no.  the mutex is release while a condition variable is waiting.